drivers/tty/serial/8250/8250_dfl.c

```diff
@@ -2,160 +2,176 @@
 /*
  * Driver for FPGA UART
  *
- * Copyright (C) 2022 Intel Corporation, Inc.
+ * Copyright (C) 2022 Intel Corporation.
  *
  * Authors:
  *   Ananda Ravuri <ananda.ravuri@intel.com>
  *   Matthew Gerlach <matthew.gerlach@linux.intel.com>
  */
 
+#include <linux/bitfield.h>
+#include <linux/device.h>
 #include <linux/dfl.h>
-#include <linux/version.h>
+#include <linux/errno.h>
+#include <linux/ioport.h>
+#include <linux/module.h>
+#include <linux/mod_devicetable.h>
+#include <linux/types.h>
+
 #include <linux/serial.h>
 #include <linux/serial_8250.h>
-#include <linux/kernel.h>
-#include <linux/module.h>
-#include <linux/bitfield.h>
-#include <linux/io-64-nonatomic-lo-hi.h>
+
+#define DFHv1_PARAM_ID_CLK_FRQ    0x2
+#define DFHv1_PARAM_ID_FIFO_LEN   0x3
+
+#define DFHv1_PARAM_ID_REG_LAYOUT	0x4
+#define DFHv1_PARAM_REG_LAYOUT_WIDTH	GENMASK_ULL(63, 32)
+#define DFHv1_PARAM_REG_LAYOUT_SHIFT	GENMASK_ULL(31, 0)
 
 struct dfl_uart {
-	void __iomem   *csr_base;
-	u64             csr_addr;
-	unsigned int    csr_size;
-	struct device  *dev;
-	u64             uart_clk;
-	u64             fifo_len;
-	unsigned int    fifo_size;
-	unsigned int    reg_shift;
-	unsigned int    line;
+	int line;
 };
 
-int feature_uart_walk(struct dfl_uart *dfluart, resource_size_t max)
+static int dfh_get_u64_param_val(struct dfl_device *dfl_dev, int param_id, u64 *pval)
 {
-	void __iomem *param_base;
-	int off;
-	u64 v;
+	size_t psize;
+	u64 *p;
 
-	v = readq(dfluart->csr_base + DFHv1_CSR_ADDR);
-	dfluart->csr_addr = FIELD_GET(DFHv1_CSR_ADDR_MASK, v);
+	p = dfh_find_param(dfl_dev, param_id, &psize);
+	if (IS_ERR(p))
+		return PTR_ERR(p);
 
-	v = readq(dfluart->csr_base + DFHv1_CSR_SIZE_GRP);
-	dfluart->csr_size = FIELD_GET(DFHv1_CSR_SIZE_GRP_SIZE, v);
-
-	if (dfluart->csr_addr == 0 || dfluart->csr_size == 0) {
-		dev_err(dfluart->dev, "FIXME bad dfh address and size\n");
+	if (psize != sizeof(*pval))
 		return -EINVAL;
-	}
 
-	if (!FIELD_GET(DFHv1_CSR_SIZE_GRP_HAS_PARAMS, v)) {
-		dev_err(dfluart->dev, "missing required parameters\n");
-		return -EINVAL;
-	}
+	*pval = *p;
 
-	param_base = dfluart->csr_base + DFHv1_PARAM_HDR;
+	return 0;
+}
 
-	off = dfl_find_param(param_base, max, DFHv1_PARAM_ID_CLK_FRQ);
-	if (off < 0) {
-		dev_err(dfluart->dev, "missing CLK_FRQ param\n");
-		return -EINVAL;
-	}
+static int dfl_uart_get_params(struct dfl_device *dfl_dev, struct uart_8250_port *uart)
+{
+	struct device *dev = &dfl_dev->dev;
+	u64 fifo_len, clk_freq, reg_layout;
+	u32 reg_width;
+	int ret;
 
-	dfluart->uart_clk = readq(param_base + off + DFHv1_PARAM_DATA);
-	dev_dbg(dfluart->dev, "UART_CLK_ID %llu Hz\n", dfluart->uart_clk);
+	ret = dfh_get_u64_param_val(dfl_dev, DFHv1_PARAM_ID_CLK_FRQ, &clk_freq);
+	if (ret) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 9, 0) && RHEL_RELEASE_CODE < 0x803
+		return dev_err(dev,
```

Comment:
On CentOS 8.2 there is compilation issue:
/home/rc/acbackport/backport/linux-dfl-backport/drivers/tty/serial/8250/8250_dfl.c:62:10: error: void value not ignored as it ought to be
   return dev_err(dev,
          ^~~~~~~~~~~~
This occurs in all the places in this file where you put in dev_err.